core/arch/arch_exports.h

```diff
@@ -1674,8 +1674,23 @@ _dynamorio_runtime_resolve(void);
 #    define APP_PARAM(mc, offs) APP_PARAM_##offs(mc)
 #endif /* X86/ARM */
 
-#define MCXT_SYSNUM_REG(mc) ((mc)->IF_X86_ELSE(xax, IF_ARM_ELSE(r7, r8)))
-#define MCXT_FIRST_REG_FIELD(mc) ((mc)->IF_X86_ELSE(xdi, r0))
+#if !(defined(MACOS) && defined(AARCH64))
+#    define MCXT_SYSNUM_REG(mc) ((mc)->IF_X86_ELSE(xax, IF_ARM_ELSE(r7, r8)))
+#    define MCXT_SET_SYSNUM_REG(mc, sysnr) \
+        ((mc)->IF_X86_ELSE(xax, IF_ARM_ELSE(r7, r8))) = (sysnr)
+#    define MCXT_FIRST_REG_FIELD(mc) &((mc)->IF_X86_ELSE(xdi, r0))
+#else
+// syscall() sets r16 to 0 and uses r0 for sysnr; libsystem_c uses r16
```

Comment:
You mean the library function syscall()?  DR goes straight to the kernel gateway if it can and doesn't try to deal with library routines.  I don't understand this though: does the kernel have multiple conventions?  Or is DR intercepting multiple library routines?